### ISSUE TITLE
False positive on enums with custom __new__

### ISSUE THREAD

**Issue by exhuma** [PyCQA/pylint#1801]
### Steps to reproduce

Lint the following code:

```python
'''
Example demonstrating a false-positive of the 'no-value-for-parameter' error
in pylint.
'''

from enum import Enum


class EnumWithDoc(Enum):
    '''
    See https://docs.python.org/3/library/enum.html#using-a-custom-new
    '''

    def __new__(cls, value, doc):
        obj = object.__new__(cls)
        obj._value_ = value  # <-- False warning reported here
        obj.doc = doc
        return obj


class Example(EnumWithDoc):
    '''
    Example enum using the "EnumWithDoc" class.
    '''

    A = (1, 'The first value')
    B = (2, 'The second value')


def main():
    '''
    Main method
    '''

    my_value = Example(1)  # <-- False error reported here!
    print(repr(my_value))
    print(my_value.doc)


main()
```

### Current behavior

The above code will return the following:

    ************* Module example
    W: 16, 8: Access to a protected member _value_ of a client class (protected-access)
    E: 35,15: No value for argument 'doc' in constructor call (no-value-for-parameter)

    ------------------------------------------------------------------
    Your code has been rated at 6.00/10 (previous run: 6.00/10, +0.00)

This indicates two problems which should not be reported:

* The warning about accessing the `_value_` attribute. When overriding `__new__`, setting this value is recommended [in the official docs](https://docs.python.org/3/library/enum.html#supported-dunder-names) ([2](https://docs.python.org/3/library/enum.html#using-a-custom-new)) and should not be reported as warning in pylint.
* The error gives the impression that the code is broken, but it isn't. This is quite a big problem as the error is reported *everyehere* where an enum value is created from a value (using `EnumClass(value)`) if that enum class happens to override `__new__`. FWIW, one of our internal libraries uses an enum close to the one above causing all of our projects to be riddled with false reports of `no-value-for-parameter`. The only way to solve this is to disable this *almost* globally. The following note from the Python docs about using custom `__new__` in enums is also relevant:

  > **Note:** The `__new__()` method, if defined, is used during creation of the Enum members; it is then replaced by Enum’s `__new__()` which is used after class creation for lookup of existing members.

### Expected behavior

No error or warning should be reported!

### pylint --version output

    pylint 1.8.1,
    astroid 1.6.0
    Python 3.5.2 (default, Nov 23 2017, 16:37:01)
    [GCC 5.4.0 20160609]

**Comment by exhuma** [comment]
I'm not sure the new title is correct. What happens is that Python will use the custom `__new__` to create enum values when the class is parsed (I suspect some metaclass magic, but have not looked that deep yet).

So, when looking at the class without executing it, it looks like:

```python
class MyEnum(Enum):

    def __new__(cls, foo, bar, baz):
        pass
```

when at actual run-time (after said Enum-magic happened) it will look like this:

```python
class MyEnum(Enum):

    def __new__(cls, value):
        ...
```

So writing the following line *at runtime* it perfectly valid:

    my_value = MyEnum(1)

But it seems that pylint sees the *originally written* class definition and expects more than one argument.

The new title suggests that the error only happens when an enum is created inside the `__new__` method.

**Comment by exhuma** [comment]
I've done some digging, and it seems that my assumption of metaclass-magic was correct. And it looks like this one could be a bit tricky. I've tracked it down to the following pieces of code:

* https://github.com/python/cpython/blob/master/Lib/enum.py#L122
* https://github.com/python/cpython/blob/master/Lib/enum.py#L461
* https://github.com/python/cpython/blob/master/Lib/enum.py#L513

The code is fairly well documented so I will not attempt to repeat the docs here, in fear of making conceptual mistakes ;)

**Comment by PCManticore** [comment]
I presume our enum astroid brain tuple will need some improvements: https://github.com/PyCQA/astroid/blob/master/astroid/brain/brain_namedtuple_enum.py#L233. I updated the title to reflect that this is not about `__new__`.

**Comment by Hawk777** [comment]
I don’t know when, but this appears to have been fixed at some point—at least, constructing an `http.HTTPStatus` by passing a status code as a parameter value used to give this error, but no longer does. Is this fixed now?

**Comment by exhuma** [comment]
I've double checked with pylint 2.4.4

```
› ./env/bin/pylint --version                                                                                                                                                                      ✗: 2
pylint 2.4.4
astroid 2.3.3
Python 3.5.2 (default, Oct  8 2019, 13:06:37)
[GCC 5.4.0 20160609]
```

and I still get the error about the "doc" argument. The warning with the `_value_` protected access is indeed gone.

**Comment by Pierre-Sassoulas** [comment]
The protected-access error was fixed in 2.14, the other issue still exists.

**Comment by wszamotula** [comment]
I've also run into the second issue here.  We are using a custom __new__ method in our Enum to associate multiple properties with each member, but when we lookup a member by value in our code we get an E1120 error.

**Comment by SiegeSailor** [comment]
I don't know if it's related, but it sounds similar. I am having a `no-member` error when I call `.value` or `.name` properties from an Enum class member, which is inherited from `aenum.Enum`:

```python
class _Enum(Enum):
    _settings_ = NoAlias

class Paths(_Enum):
    SECRETS = os.getenv("LIBDEV_PATH_SECRETS")

logging.info(Paths.SECRETS.value) # E1101: Instance of 'str' has no 'value' member (no-member)
```
